### DISCUSSION TITLE
Charge price calculation

### DISCUSSION THREAD

**Discussion by fabiomaticus** [flobz/psa_car_controller#249] [Q&A]
How can I change it?

**Comment by JHCD** [comment]
In your config directory there is a `config.ini`, there you can define the costs.

**Comment by fabiomaticus** [comment]
In my config directory (where is configuration.yaml) i don't have config.ini

**Reply by JHCD** [reply]
That is strange, I never saw a yaml. My config-Folder:

![grafik](https://user-images.githubusercontent.com/12496340/138092704-a5d5990d-4a72-4ac0-a762-5b5b53e3044a.png)

**Reply by fabiomaticus** [reply]
I have home assistant on Raspberry.

**Reply by JHCD** [reply]
That could be the difference, I don't know the structure of that yml...

The content of the config.ini is:
```
[General]
currency = €

[Electricity config]
day price = 0.1234
night price = 
night hour start = 
night hour end = 
```

**Comment by nuiler** [comment]
Where can I find the config file under home assistant?

**Reply by Flodu31** [reply]
You need to connect to the host, in SSH, on port 22222. After, it should be in /mnt/docker/...

**Reply by codeclinic** [reply]
IS there anyway that these values could be set in the Addon config page in future releases? I've tried everything to get 2222 access and it'S just not happening.

**Reply by Ilpo55** [reply]
I was able to edit the config.ini in the docker, but still the price calculation is wrong.

This is how i did it with Web Terminal
Disable protection mode in SSH & Web Terminal.
Find the name of the PSA Car Controller docker container.
**docker ps**
I got "addon_b9f22d83_psacc".
Run "docker exec ..." command to use the container.
**docker exec -t -i addon_b9f22d83_psacc /bin/bash**
Install nano editor in to the container.
**apt-get update**
**apt-get -y install nano**
**cd /data**
**nano config.ini**
Exit docker and restart PSA Car Controller.

**Reply by Ilpo55** [reply]
Now my charge calculations are correct.
I think that the reason for wrong calculation was that "dc charge price" was used for all charges and "dc charge price" was empty.

Now I have dc charge price = 0.10 in file config.ini

(The correct prices are with the new charges after editing "dc charge price = 0.10")

**Reply by Voz-DeCarabas** [reply]
> I was able to edit the config.ini in the docker, but still the price calculation is wrong.
> 
> This is how i did it with Web Terminal Disable protection mode in SSH & Web Terminal. Find the name of the PSA Car Controller docker container. **docker ps** I got "addon_b9f22d83_psacc". Run "docker exec ..." command to use the container. **docker exec -t -i addon_b9f22d83_psacc /bin/bash** Install nano editor in to the container. **apt-get update** **apt-get -y install nano** **cd /data** **nano config.ini** Exit docker and restart PSA Car Controller.

Thanks very much for these instructions.  I've been able to use them to update the API keys and the price, then restart the service.  My challenge is that I want to change the currency from euro (which I don't appear to have a key for) to £.  In nano it's displaying the euro symbol as ^b and when I tried replacing that with £ I can't see an entry and the PSA Car Controller would not start.  I've needed to remove, reboot and readd the add-on and start again.  cat and more of config.ini show the euro symbol ok.  Is there an easy way to make that change?  I would guess it's some character map issue but I have no idea.

**Reply by Ilpo55** [reply]
Does it help if you use  "ALT+ascii code"  method to get € or £ ?
Press and hold the ALT key and type the number 0163 to make a Pound symbol.
Press and hold the ALT key and type the number 0128 to make a Euro symbol. 
(You have to use numeric keybad to enter the numbers).

**Reply by Voz-DeCarabas** [reply]
Interesting.  The ALT + ascii  code works in the first SSH connection but doesn't display anything once in the docker container.  Using nano I changed the file, removed the old ^b and using the alt+asci to add a £, although it wasn't displayed.  Saving and then using either cat or more I could then see the £ and the file looked correct.  I restarted PSA and it wouldn't restart as below.
I'll keep playing for a bit but maybe I need to accept changes for cost and APIs but live with the wrong currency symbol.  A quick search shows a few issues reported on this around setting environmant variables for the docker, so I need to do a bit more reading.

`Containerised psa_car_controller loading...
Traceback (most recent call last):
  File "/usr/local/bin/psa-car-controller", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/dist-packages/psa_car_controller/__main__.py", line 19, in main
    app = PSACarController()
  File "/usr/local/lib/python3.9/dist-packages/psa_car_controller/psacc/utils/utils.py", line 37, in __call__
    cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/psa_car_controller/psacc/application/car_controller.py", line 54, in __init__
    self.config = ConfigRepository.read_config()
  File "/usr/local/lib/python3.9/dist-packages/psa_car_controller/psacc/repository/config_repository.py", line 162, in read_config
    config_str = ConfigRepository._read_file(name)
  File "/usr/local/lib/python3.9/dist-packages/psa_car_controller/psacc/repository/config_repository.py", line 157, in _read_file
    return f.read()
  File "/usr/lib/python3.9/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xac in position 23: invalid start byte`

**Reply by codeclinic** [reply]
> Thanks very much for these instructions. I've been able to use them to update the API keys and the price, then restart the service. My challenge is that I want to change the currency from euro (which I don't appear to have a key for) to £. In nano it's displaying the euro symbol as ^b and when I tried replacing that with £ I can't see an entry and the PSA Car Controller would not start. I've needed to remove, reboot and readd the add-on and start again. cat and more of config.ini show the euro symbol ok. Is there an easy way to make that change? I would guess it's some character map issue but I have no idea.

I have the same issue. What is this ^b ???

**Reply by codeclinic** [reply]
For what it's worth my ^b results in the Euro sign being displayed. Weird.

**Reply by Ilpo55** [reply]
€ in hex is E282AC
£ in hex is C2A3
look at https://www.online-toolz.com/tools/text-hex-convertor.php

Does it help if you use hexedit to replace the € with £ in config.ini.
```
apt-get update
apt-get install -y hexedit
cd /data
hexedit config.ini
```
replace E2 82 AC with C2 A3 20   (last 20 is space in hex, just to replace the extra error cousing AC with something)

![e_to_p_1](https://github.com/flobz/psa_car_controller/assets/120249692/23f2435d-ad4c-4143-afae-340d2195f357)
![e_to_p_2](https://github.com/flobz/psa_car_controller/assets/120249692/c32b1bec-593e-4e61-9718-3713bfb14ac7)

**Reply by Voz-DeCarabas** [reply]
Yeah.  I don't know enough about docker etc to be confident in playing around in there and everytime I break the add-in and re-install I loose my previous data, which starts to get annoying.  I 'think' it might be possible to change the location evironment settings for the Docker but I'm not confident that it wouldn't get overwritten with the next upgrade of the add-in.  More reading to do.

**Reply by Voz-DeCarabas** [reply]
> € in hex is E282AC £ in hex is C2A3 look at https://www.online-toolz.com/tools/text-hex-convertor.php
> 
> Does it help if you use hexedit to replace the € with £ in config.ini.
> 
> ```
> apt-get update
> apt-get install -y hexedit
> cd /data
> hexedit config.ini
> ```
> 
> replace E2 82 AC with C3 A3 20 (last 20 is space in hex, just to replace the extra error cousing AC with something)
> 
> ![e_to_p_1](https://user-images.githubusercontent.com/120249692/239558353-23f2435d-ad4c-4143-afae-340d2195f357.jpg) ![e_to_p_2](https://user-images.githubusercontent.com/120249692/239558378-c32b1bec-593e-4e61-9718-3713bfb14ac7.jpg)

ooooh.  that's an intersting idea.  I'll give that a go in a bit.  so much for not risking my data!

**Reply by Ilpo55** [reply]
Note: replace E2 82 AC with **C2** A3 20 ....
Pictures are correct. In the earlier text there was my error.

**Reply by Voz-DeCarabas** [reply]
Success!  Thanks for that idea!

![PSA Car Controller – Home Assistant](https://github.com/flobz/psa_car_controller/assets/23322189/ee918355-6d91-4733-832b-0826ca6ee39e)

The summary page is still showing euro symbol but there's no data in there yet, I think I need to do a few more trips.
Now I can add the correct price in the .ini and the API keys into the config.json and I should be good to go!

Thanks again.

**Reply by Ilpo55** [reply]
Great!
Maybe deleteing  and rewriteing the first 3 lines in config.ini would also have corrected the issue.
The issue was that nano did not delete <hex:AC> when you deleted the ^b.

**Reply by Voz-DeCarabas** [reply]
I've now done a trip long enough to register the stats and I can see that the Euro symbol isn't updated on the Summary tab, even though it's now right on the Charge tab.  Also, as reported by others, the date format isn't right on the charts.  Could this all be down to localisation issues in the docker container?
![PSA](https://github.com/flobz/psa_car_controller/assets/23322189/8bbeb426-4d90-4f14-92a4-058822cbd661)

**Reply by Ilpo55** [reply]
In docker, file /usr/local/lib/python3.9/dist-packages/psa_car_controller/web/figures.py is € at line 37
In GitHub psa_car_controller/web/figures.py

I am not brave enough to edit the file and see what happens.

**Reply by Voz-DeCarabas** [reply]
So sounds like a need to have the config file entry used instead and it wouldn't impact the date format.  I also see that the instructions have been updated for the HA add-on https://github.com/flobz/psacc-ha/tree/main/psacc-ha#modify-config-file for backing up the database so I might have a go at changing the hard coded value in the deployed code and seeing what happens (hex editor time again I assume).

**Reply by Voz-DeCarabas** [reply]
> In docker, file /usr/local/lib/python3.9/dist-packages/psa_car_controller/web/figures.py is € at line 37 In GitHub psa_car_controller/web/figures.py
>
Looking in the file it also occured to me that it's hard coded for km's so another enhancement would be to have the units selectable for miles.  Interestingly the select statement reads 'mileage' from the db not 'kilometerage' 😆

**Reply by Ilpo55** [reply]
> I also see that the instructions have been updated for the HA add-on 

Did you succeed to edit the files with the new instructions?

I get the error "no such file or directory"
```
➜  ~ cd /mnt/data/supervisor/addons/data/b9f12d83_psacc
cd: no such file or directory: /mnt/data/supervisor/addons/data/b9f12d83_psacc
➜  ~
```

**Reply by Voz-DeCarabas** [reply]
I hadn't tried, as I prefer your instructions.  
I just did and same as you it returns a path that doesn't exist (the /mnt directory is empty).

**Reply by Ilpo55** [reply]
Thanks,
I will add comment to https://github.com/Flodu31/HomeAssistant-PeugeotIntegration/issues/34#issuecomment-1558632856

**Comment by wifi75** [comment]
in home assistant i can't find the config.ini file.
but where can i find it?

**Reply by Ilpo55** [reply]
Did you try with my (Feb 22) instructions?

**Reply by codeclinic** [reply]
1. cd /data
2. nano config.ini

**Comment by codeclinic** [comment]
I found the easiest way to do this is to install the portainer addon, then follow the same instructions as above.

**Comment by Strikky** [comment]


**Reply by codeclinic** [reply]
Dude, you won't get much help with that attitude.

Good luck!

**Reply by codeclinic** [reply]
There you go again. You do realise no-one is obliged to help you... right?

**Reply by wormvortex** [reply]
Is there a way to display this charge price data inside home assistant?
